cmd/load-generator/challenge-servers.go

```diff
@@ -0,0 +1,151 @@
+package main
+
+import (
+	rand "crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"math/big"
+	"net/http"
+	"os"
+	"strings"
+	"sync"
+	"time"
+)
+
+type challSrv struct {
+	hoMu        sync.RWMutex
+	httpOne     map[string]string
+	httpOneAddr string
+
+	tlsOneAddr string
+
+	// doMu       sync.RWMutex
+	// dnsOne     map[string]string
+	// dnsOneAddr string
+}
+
+func newChallSrv(httpOneAddr, tlsOneAddr string) *challSrv {
+	return &challSrv{
+		httpOne:     make(map[string]string),
+		httpOneAddr: httpOneAddr,
+		tlsOneAddr:  tlsOneAddr,
+	}
+}
+
+// Run runs the challenge server on the configured address
+func (s *challSrv) run() {
+	wg := new(sync.WaitGroup)
+	wg.Add(1)
+	go func() {
+		err := s.httpOneServer(wg)
+		if err != nil {
+			fmt.Printf("[!] http-0 server failed: %s\n", err)
+			os.Exit(1)
+		}
+	}()
+	wg.Add(1)
+	go func() {
+		err := s.tlsOneServer(wg)
+		if err != nil {
+			fmt.Printf("[!] tls-sni-01 server failed: %s\n", err)
+			os.Exit(1)
+		}
+	}()
+	wg.Wait()
+}
+
+func (s *challSrv) addHTTPOneChallenge(token, content string) {
+	s.hoMu.Lock()
+	defer s.hoMu.Unlock()
+	s.httpOne[token] = content
+}
+
+func (s *challSrv) deleteHTTPOneChallenge(token string) {
+	s.hoMu.Lock()
+	defer s.hoMu.Unlock()
+	if _, ok := s.httpOne[token]; ok {
+		delete(s.httpOne, token)
+	}
+}
+
+func (s *challSrv) getHTTPOneChallenge(token string) (string, bool) {
+	s.hoMu.RLock()
+	defer s.hoMu.RUnlock()
+	content, present := s.httpOne[token]
+	return content, present
+}
+
+func (s *challSrv) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	requestPath := r.URL.Path
+	if strings.HasPrefix(requestPath, "/.well-known/acme-challenge/") {
+		token := requestPath[28:]
```

Comment:
once the path prefix is a const it would probably make sense to change `28:` to `len(theConst):` or equiv. Less magic looking :crystal_ball: